Add support for compiling on windows x64 and x86 with vcpkg-provided dependencies

talregev opened the pull request:
Compile for windows x64 and x86.
Upload the artifacts. You can download and test it yourself. 
x64 not running.
~~x86 only splash appear:~~
x86 full running!
![image](https://github.com/gazebosim/gazebo-classic/assets/7043539/7d037a3e-841a-459c-af49-53db3a5637c8)


@traversaro
I know I get full ignore because I open an issue telling that here we get a low responds from maintainers.  
Even with your ignore, I am still right, and I am still keep develop for gazebo classic.
This things that I bring here, can done also for the new gazebo with the help of the community.

traversaro commented:
> I know I get full ignore because I open an issue telling that here we get a low responds from maintainers.

Actually I had just lost this notification.

talregev committed:
Compile for windows x64 and x86

talregev commented:
@traversaro 
Can I do anything to help getting this PR?

talregev commented:
the gzserver.exe seems to work. When I open it separately it run without crash, then:
gz topic -l 
![image](https://github.com/gazebosim/gazebo-classic/assets/7043539/fb7bba66-f498-4680-ba57-997d3ebfd52a)


So it compile correctly.

talregev commented:
I was able to run gzclient in x86!
I did gzclient.exe --verbose to show what it missing.
It was missing this folder: `C:\Users\<my user name>\.gazebo\`
And missing HOME variable
and I create HOME variable to:
`set HOME=C:\Users\<my user name>`

EDIT: The same way I was able to run gazebo.exe

Then it show up!
![image](https://github.com/gazebosim/gazebo-classic/assets/7043539/43b233a5-9d7c-4b8a-adec-80562174b9b1)

talregev commented:
@traversaro 
For windows it shouldn't search for HOME variable. it should search for USERPROFILE.
Also it should create automatic this folder:
`C:\Users\<my user name>\.gazebo\`

talregev commented:
![image](https://github.com/gazebosim/gazebo-classic/assets/7043539/a8a0eacc-78fb-429d-8a68-7c7663afee28)

talregev commented:
@traversaro How I can help to move this PR forward? It already running.

talregev commented:
@scpeter can you take a look?

talregev commented:
@scpeters Can you take a look?

talregev commented:
@scpeters Can you test locally that I didn't break the regular gazebo classic build?

talregev commented:
@traversaro @scpeters 
How I can help to move this PR forward?

talregev commented:
@traversaro Thank you for the merge.
Can you take a look on this PR also?

talregev commented:
@traversaro Any reason to block this PR?
We can discuss and change stuff to move this forward.
Is it something you want that you can compile gazebo without conda?

traversaro commented:
> @traversaro Any reason to block this PR? We can discuss and change stuff to move this forward. Is it something you want that you can compile gazebo without conda?

Sorry for the late reply, I hve some doubts about the mantainability of pinning ogre, qwt and graphviz without pinning the other dependencies. Anyhow, we can discuss about this with the other mantainers. In the meanwhile I can review the rest of the PR, that probably can also merged without the vcpkg.json and related CI.

traversaro commented:
Just to understand, x64 is still not running, right?

traversaro commented on .github/workflows/windows.yml:
```diff
@@ -0,0 +1,109 @@
+name: 🪟 Windows
```
Can you rename this workflow and related files from `windows`  to `windows-vcpkg`  or similar? Compilation for Windows can be achieved with other package managers (like conda or conan), so marking this as `windows`  may be a bit confusing.

talregev replied:
I will rename it.

talregev replied:
done.

talregev commented:
The x64 is still not working.
I can remove it from running on the ci if you wish.

talregev commented:
> > @traversaro Any reason to block this PR? We can discuss and change stuff to move this forward. Is it something you want that you can compile gazebo without conda?
> 
> Sorry for the late reply, I hve some doubts about the mantainability of pinning ogre, qwt and graphviz without pinning the other dependencies. Anyhow, we can discuss about this with the other mantainers. In the meanwhile I can review the rest of the PR, that probably can also merged without the vcpkg.json and related CI.

All the dependencies are already pinned.
They pinned to a specific commit of vcpkg. 
It means they do not get the latest update and stay the same always.

traversaro commented on cmake/FindFreeimage.cmake:
```diff
@@ -4,7 +4,14 @@ include (${gazebo_cmake_dir}/GazeboUtils.cmake)
 # Find packages
 if (PKG_CONFIG_FOUND)
-  pkg_check_modules(freeimage freeimage>=${MIN_FREEIMAGE_VERSION})
+  find_package(freeimage CONFIG)
+  if (freeimage_FOUND)
+    set(freeimage_LIBRARIES freeimage::FreeImage)
+    message (STATUS "FreeImage found")
+  endif()
```
Can you move this block after the `endif (PKG_CONFIG_FOUND)` block? In this way we ensure that nothing at all changes for setups that find freeimage with pkg-config, and it is also more logical that we call `find_package(freeimage CONFIG)`  even if pkg-config is not installed.

talregev replied:
I will try.

talregev replied:
done.

traversaro commented on cmake/SearchForStuff.cmake:
```diff
@@ -164,9 +164,12 @@ if (PKG_CONFIG_FOUND)
-    message (STATUS "Looking for Simbody - found")
+  find_package(Simbody CONFIG)
+  if (Simbody_FOUND)  
+    if ("${Simbody_LIBRARIES}" STREQUAL "" OR "${Simbody_LIBRARIES}" STREQUAL "Simbody_LIBRARIES-NOTFOUND")
```
Can you add a comment here explaining why/when this workaround is necessary? Thanks

talregev replied:
Done.

traversaro commented on cmake/SearchForStuff.cmake:
```diff
@@ -164,9 +164,12 @@ if (PKG_CONFIG_FOUND)
-  find_package(Simbody)
-  if (Simbody_FOUND)
-    message (STATUS "Looking for Simbody - found")
+  find_package(Simbody CONFIG)
```
Unless someone is shipping a `FindSimbody.cmake` and putting it in the `CMAKE_MODULE_PATH`, I guess there is no reason to explicitly ask for `CONFIG`,right?
```suggestion
  find_package(Simbody)
```

talregev replied:
This is how it done for vcpkg.
I think it not find it without config.

talregev replied:
I will try.

talregev replied:
The vcpkg give the correct parameters to find it with config.
I will try without.

traversaro replied:
Unless a `FindSimbody.cmake` script is present somehow in `CMAKE_MODULE_PATH`, `find_package(Simbody)` should just fallback to `find_package(Simbody CONFIG)`, see https://cmake.org/cmake/help/latest/command/find_package.html for more info. vcpkg suggests `find_package(<pkg> CONFIG)` in its install message, but `find_package(<pkg>)` should work fine as well. For this reason, I would prefer to minimize the unneccessary changes and for consistency with the other dependencies, thanks!

talregev replied:
ok. I already trying all your suggestion. Something is not working, and I am not sure what.

talregev replied:
done.

traversaro commented on cmake/SearchForStuff.cmake:
```diff
@@ -164,9 +164,12 @@ if (PKG_CONFIG_FOUND)
+    if ("${Simbody_LIBRARIES}" STREQUAL "" OR "${Simbody_LIBRARIES}" STREQUAL "Simbody_LIBRARIES-NOTFOUND")
+      set(Simbody_LIBRARIES SimTKmath SimTKcommon SimTKsimbody)
+    endif()
+    message (STATUS "Simbody_LIBRARIES: ${Simbody_LIBRARIES}")
```
This seems a debug message, can we remove it?
```suggestion
```

talregev replied:
done.

talregev commented:
vcpkg.json is part of the build and part of the compilation process.
It is the same to tell me, I can approve all the cpp and header files but the cmake files we leave behind.

traversaro commented on cmake/SearchForStuff.cmake:
```diff
@@ -213,7 +216,13 @@ if (PKG_CONFIG_FOUND)
+    endif()
+    if (NOT tinyxml_FOUND)
+      pkg_check_modules(tinyxml tinyxml)
+    endif()
```
For backward compatibility, can we search via `find_package(tinyxml CONFIG)`  only if `    pkg_check_modules(tinyxml tinyxml)` fails?

talregev replied:
I will try.

talregev replied:
Done.

traversaro commented on cmake/SearchForStuff.cmake:
```diff
@@ -327,7 +336,14 @@ if (PKG_CONFIG_FOUND)
   #################################################
   # Find TBB
-  pkg_check_modules(TBB tbb)
+  find_package(TBB CONFIG REQUIRED)
```
For backward compatibility, can we search via `find_package(TBB CONFIG)`  only if `    pkg_check_modules(TBB tbb)` fails?

talregev replied:
I will try.

talregev replied:
It make errors. Can we try different solution?

traversaro commented on cmake/SearchForStuff.cmake:
```diff
@@ -355,6 +371,7 @@ if (PKG_CONFIG_FOUND)
   if (DEFINED TBB_VERSION AND NOT ${TBB_VERSION} STREQUAL "")
     if (${TBB_VERSION} VERSION_GREATER_EQUAL "2021.0")
       set(HAVE_TBB_GREATER_OR_EQUAL_2021 ON)
+      message (STATUS "HAVE_TBB_GREATER_OR_EQUAL_2021=ON")
```
This seems a debug message, can we remove it?

talregev replied:
I will remove it.

talregev replied:
done.

traversaro commented on cmake/SearchForStuff.cmake:
```diff
@@ -583,13 +600,20 @@ if (PKG_CONFIG_FOUND)
   # so check it first. if it is not present, check for the OSRF
   # custom bullet2.82.pc file
-  pkg_check_modules(BULLET bullet>=2.82)
+  find_package(BULLET CONFIG 2.82)
```
For backward compatibility, can we search via `find_package`  only if `    pkg_check_modules` fails?

talregev replied:
I will try.

talregev replied:
done.

traversaro commented on cmake/SearchForStuff.cmake:
```diff
@@ -635,7 +659,7 @@ endif ()
 # Find SDFormat
 set(SDF_MIN_REQUIRED_VERSION 9.8)
-find_package(sdformat9 ${SDF_MIN_REQUIRED_VERSION} REQUIRED)
+find_package(sdformat9 CONFIG ${SDF_MIN_REQUIRED_VERSION} REQUIRED)
```
Why it is necessary to add `CONFIG`  here?

talregev replied:
sdformat9 also come from vcpkg. This is how vcpkg search. It search with config. 
I can try without config, but I am not sure how if it work.
The config also work for conda. 
Are you insist on it?

talregev replied:
I will try to remove it.

talregev replied:
done.

traversaro commented on cmake/SearchForStuff.cmake:
```diff
@@ -679,13 +703,18 @@ endif ()
   set (HAVE_GDAL OFF CACHE BOOL "HAVE GDAL" FORCE)
 else ()
   message (STATUS "Looking for libgdal - found")
+  message (STATUS "GDAL_LIBRARY: ${GDAL_LIBRARY}")  
```
Debug message, perhaps we can remove it?

talregev replied:
We can remove debug msg.

talregev replied:
done.

traversaro commented on gazebo/rendering/RenderEngine.cc:
```diff
@@ -22,6 +22,8 @@
 
 // Not Apple or Windows
 #if not defined(__APPLE__) && not defined(_WIN32)
+#define GL_GLEXT_LEGACY
+#define GLX_GLXEXT_LEGACY
```
This code is only related to Linux, why it is needed in this PR that from what I understand is Windows only?

talregev replied:
I will try to remove it.

talregev replied:
done.

traversaro commented:
> They pinned to a specific commit of vcpkg.

I probably missed at which point we are pinning a specific commit/version of vcpkg, can you point it to me? Thanks!

talregev commented:
> > They pinned to a specific commit of vcpkg.
> 
> I probably missed at which point we are pinning a specific commit/version of vcpkg, can you point it to me? Thanks!

It inside the json:

https://github.com/gazebosim/gazebo-classic/pull/3320/files#diff-9a0ed7d1415a6669e62312071492f88570d14b5d75b59557bbce9dfc2ab7b933R6
```json
  "builtin-baseline": "19af97cba8ca48474e4ad15a24ed50271a9ecdac",
```

talregev committed:
Refactor according comments.

traversaro commented:
> > > They pinned to a specific commit of vcpkg.
> > 
> > 
> > I probably missed at which point we are pinning a specific commit/version of vcpkg, can you point it to me? Thanks!
> 
> It inside the json:
> 
> https://github.com/gazebosim/gazebo-classic/pull/3320/files#diff-9a0ed7d1415a6669e62312071492f88570d14b5d75b59557bbce9dfc2ab7b933R6
> 
> ```json
>   "builtin-baseline": "19af97cba8ca48474e4ad15a24ed50271a9ecdac",
> ```

Cool, thanks!

talregev commented:
@traversaro Your suggestion for make the cmake back compatible before the changes, make the ci and builds to have errors.
Before your changes, both the old builds and the new builds was working.
Can you help me to find where is your problem on your changes? Also can you reconsider my changes?
currently I brute force search what cause the errors. 
Thank you in advanced.

traversaro commented:
> @traversaro Your suggestion for make the cmake back compatible before the changes, make the ci and builds to have errors. Before your changes, both the old builds and the new builds was working. Can you help me to find where is your problem on your changes? Also can you reconsider my changes? currently I brute force search what cause the errors. Thank you in advanced.

Beside people that consume gazebo from official binaries (apt and homebrew) many users compile Gazebo Classic from sources in their environments/distributions with their own way of installing dependencies. If Gazebo Classic was an active project, I would be happy to cleanup how we find CMake dependencies in a new major version. However, Gazebo Classic is a project in mantainance mode, so I think we should try to minimize the risk of regressions, especially if we are not sure why this is necessary. 

Can you report the error that occur if we search first with `pkg_check_modules`? In theory pkg-config should work just fine on Windows/with vcpkg dependencies. From the CI it just seems that the CI runs out of time.

talregev commented:
> > @traversaro Your suggestion for make the cmake back compatible before the changes, make the ci and builds to have errors. Before your changes, both the old builds and the new builds was working. Can you help me to find where is your problem on your changes? Also can you reconsider my changes? currently I brute force search what cause the errors. Thank you in advanced.
> 
> Beside people that consume gazebo from official binaries (apt and homebrew) many users compile Gazebo Classic from sources in their environments/distributions with their own way of installing dependencies. If Gazebo Classic was an active project, I would be happy to cleanup how we find CMake dependencies in a new major version. However, Gazebo Classic is a project in mantainance mode, so I think we should try to minimize the risk of regressions, especially if we are not sure why this is necessary.
> 
> Can you report the error that occur if we search first with `pkg_check_modules`? In theory pkg-config should work just fine on Windows/with vcpkg dependencies. From the CI it just seems that the CI runs out of time.

I am not talking about vcpkg ci. I am talking conda-windows ci. It missing something and run the protobuf command with missing command, and only with parameters. It fail on the ci.

traversaro commented:
> I am not talking about vcpkg ci. I am talking conda-windows ci. It missing something and run the protobuf command with missing command, and only with parameters. It fail on the ci.

This is unrelated to the changes in your PR, see https://github.com/gazebosim/gazebo-classic/issues/3339 . I am working on this, but it is not related to this PR.

talregev commented:
> > I am not talking about vcpkg ci. I am talking conda-windows ci. It missing something and run the protobuf command with missing command, and only with parameters. It fail on the ci.
> 
> This is unrelated to the changes in your PR, see #3339 . I am working on this, but it is not related to this PR.

Are you sure?? Because if I will revert your suggestion it will work.
~~Also it working on the main branch. without any changes.~~

I can re-run the CI with your changes again only to check vcpkg ci. I will ignore the conda-windows ci.

traversaro commented:
> Are you sure?? Because if I will revert your suggestion it will work.

The last successful build for that CI job is from a week ago, see https://github.com/gazebosim/gazebo-classic/actions/workflows/conda-forge.yml .

talregev commented:
> > Are you sure?? Because if I will revert your suggestion it will work.
> 
> The last successful build for that CI job is from a week ago, see https://github.com/gazebosim/gazebo-classic/actions/workflows/conda-forge.yml .

Yes, I understand. Thank you for your explanations.

talregev commented:
@traversaro 
I am reverting the changes for tbb:
If you have another idea, I will happy to hear.

```
-- Checking for module 'tbb'
--   Package 'tbb', required by 'virtual:world', not found
CMake Error at vcpkg/scripts/buildsystems/vcpkg.cmake:855 (_find_package):
  Found package configuration file:

    D:/a/gazebo-classic/gazebo-classic/vcpkg/installed/x86-windows-release/share/tbb/TBBConfig.cmake

  but it set TBB_FOUND to FALSE so package "TBB" is considered to be NOT
  FOUND.
Call Stack (most recent call first):
  cmake/SearchForStuff.cmake:340 (find_package)
  CMakeLists.txt:163 (include)


-- Configuring incomplete, errors occurred!
```

talregev committed:
revert changes of tbb

talregev commented:
@traversaro The build is pass with your suggestions without the changes for tbb.
How you would like to continue?

talregev committed:
simbody workaround comment explanation.

talregev commented:
@traversaro I answer all your comments review except the tbb changes. 
Your suggestion make the build to fail. Any suggestion how to solve it and also be backward compatible as you want? 
Thank you for your help.

traversaro commented:
> @traversaro I am reverting the changes for tbb: If you have another idea, I will happy to hear.
> 
> ```
> -- Checking for module 'tbb'
> --   Package 'tbb', required by 'virtual:world', not found
> CMake Error at vcpkg/scripts/buildsystems/vcpkg.cmake:855 (_find_package):
>   Found package configuration file:
> 
>     D:/a/gazebo-classic/gazebo-classic/vcpkg/installed/x86-windows-release/share/tbb/TBBConfig.cmake
> 
>   but it set TBB_FOUND to FALSE so package "TBB" is considered to be NOT
>   FOUND.
> Call Stack (most recent call first):
>   cmake/SearchForStuff.cmake:340 (find_package)
>   CMakeLists.txt:163 (include)
> 
> 
> -- Configuring incomplete, errors occurred!
> ```

Can you point me to the version of the code that is giving this error? I suspect this is related to https://gitlab.kitware.com/cmake/cmake/issues/17135, but I can't say for sure unless I check the code.

(In general, please do not ping three times in a few hours if nobody replies, sometime it takes time to reply).

traversaro commented:
> Can you point me to the version of the code that is giving this error? I suspect this is related to https://gitlab.kitware.com/cmake/cmake/issues/17135, but I can't say for sure unless I check the code.

In particular, why did you added a `find_package(TBB)` before the `unset(TBB_FOUND CACHE)` later in the code in https://github.com/gazebosim/gazebo-classic/pull/3320/commits/bd3c45c18f2138260cd2a0405c9e0a6fe98d7a18#diff-070f9711d42593f839e9cb29c2bcb31f999939af11125c208965891a2ba1617cR340 ? The `unset(TBB_FOUND CACHE)` is exactly what should avoid the error you are seeing, see https://github.com/gazebosim/gazebo-classic/commit/183122dc0d6f96d030c48b1a09d3d46d962fad45 for a related commit.

traversaro commented on cmake/SearchForStuff.cmake:
```diff
@@ -165,8 +165,11 @@ if (PKG_CONFIG_FOUND)
+  if (Simbody_FOUND)  
+    # When simbody is found but it don't define any libs, we should manually define the libs.
+    if ("${Simbody_LIBRARIES}" STREQUAL "" OR "${Simbody_LIBRARIES}" STREQUAL "Simbody_LIBRARIES-NOTFOUND")
+      set(Simbody_LIBRARIES SimTKmath SimTKcommon SimTKsimbody)
```
From what I see from https://github.com/simbody/simbody/blob/252dfb5957b853e181555c63a546854a57ad267e/cmake/SimbodyConfig.cmake.in#L97, `Simbody_LIBRARIES` is not defined only when shared libraries are not build and instead static libraries are built. However, in that case the libraries are not called  SimTKmath SimTKcommon SimTKsimbody, but rather  SimTKmath_static SimTKcommon_static SimTKsimbody_static . So I either change this line to `set(Simbody_LIBRARIES ${Simbody_STATIC_LIBRARIES})`, or even better if possible I would report here the `SimbodyConfig.cmake` in the case `Simbody_LIBRARIES` is not defined.

talregev replied:
I will revert the changes of tbb, and it should work as you say. I didn't notice I added twice find package tbb.

talregev replied:
How to report the SimbodyConfig.cmake?

talregev replied:
I choose the first option.

traversaro commented on cmake/SearchForStuff.cmake:
```diff
@@ -214,6 +217,12 @@ if (PKG_CONFIG_FOUND)
     message (STATUS "Using system tinyxml.")
     pkg_check_modules(tinyxml tinyxml)
+    if (NOT tinyxml_FOUND)
+      find_package(tinyxml CONFIG)
```
Are you sure that this code is actually useful? From the config log, I see:
~~~
2023-08-16T15:47:18.9015616Z -- Checking for module 'tinyxml'
2023-08-16T15:47:18.9368445Z --   Package 'tinyxml', required by 'virtual:world', not found
2023-08-16T15:47:18.9433443Z CMake Warning at vcpkg/scripts/buildsystems/vcpkg.cmake:855 (_find_package):
2023-08-16T15:47:18.9434230Z   Found package configuration file:
2023-08-16T15:47:18.9434627Z 
2023-08-16T15:47:18.9435374Z     D:/a/gazebo-classic/gazebo-classic/vcpkg/installed/x86-windows-release/share/tinyxml/tinyxml-config.cmake
2023-08-16T15:47:18.9436143Z 
2023-08-16T15:47:18.9436475Z   but it set tinyxml_FOUND to FALSE so package "tinyxml" is considered to be
2023-08-16T15:47:18.9437060Z   NOT FOUND.
2023-08-16T15:47:18.9437363Z -- Using system tinyxml2.
2023-08-16T15:47:18.9438008Z Call Stack (most recent call first):
2023-08-16T15:47:18.9438504Z   cmake/SearchForStuff.cmake:221 (find_package)
2023-08-16T15:47:18.9438974Z   CMakeLists.txt:163 (include)
2023-08-16T15:47:18.9439304Z 
2023-08-16T15:47:18.9439420Z 
2023-08-16T15:47:18.9454750Z -- Checking for module 'tinyxml2'
2023-08-16T15:47:19.0120119Z --   Found tinyxml2, version 9.0.0
~~~

The reason for this error is basically the same of tbb, i.e. https://gitlab.kitware.com/cmake/cmake/issues/17135 . You need to unset `tinyxml_FOUND` after `if (NOT tinyxml_FOUND)`. However, as no error message is printed, probably tinyxml is found correctly by the find_path and find_library calls. later on. If everything is already working, perhaps we can drop all the tinyxml related changes?

talregev replied:
I prefer to do it in the right way. I can unset the found like tbb.

traversaro replied:
There is any advantage in the `find_package(tinyxml)` way? If no, please avoid making changes that could create regression in other (untested) situations, thanks.

talregev replied:
I revert the changes.

talregev commented:
> > Can you point me to the version of the code that is giving this error? I suspect this is related to https://gitlab.kitware.com/cmake/cmake/issues/17135, but I can't say for sure unless I check the code.
> 
> In particular, why did you added a `find_package(TBB)` before the `unset(TBB_FOUND CACHE)` later in the code in [bd3c45c#diff-070f9711d42593f839e9cb29c2bcb31f999939af11125c208965891a2ba1617cR340](https://github.com/gazebosim/gazebo-classic/commit/bd3c45c18f2138260cd2a0405c9e0a6fe98d7a18#diff-070f9711d42593f839e9cb29c2bcb31f999939af11125c208965891a2ba1617cR340) ? The `unset(TBB_FOUND CACHE)` is exactly what should avoid the error you are seeing, see [183122d](https://github.com/gazebosim/gazebo-classic/commit/183122dc0d6f96d030c48b1a09d3d46d962fad45) for a related commit.

I will revert the changes of tbb, and it should work as you say. I didn't notice I added twice find package tbb.

traversaro commented on cmake/FindFreeimage.cmake:
```diff
@@ -9,7 +9,13 @@ if (PKG_CONFIG_FOUND)
+    set(freeimage_LIBRARIES freeimage::FreeImage)
+    message (STATUS "FreeImage found")
+  endif()
+endif()
```
As for tinyxml, this `find_package(freeimage CONFIG)` is not working, but freeimage is found by the fallback logic:
~~~
2023-08-16T15:47:12.6019654Z CMake Warning at vcpkg/scripts/buildsystems/vcpkg.cmake:855 (_find_package):
2023-08-16T15:47:12.6020579Z   Found package configuration file:
2023-08-16T15:47:12.6021019Z 
2023-08-16T15:47:12.6021899Z     D:/a/gazebo-classic/gazebo-classic/vcpkg/installed/x86-windows-release/share/freeimage/freeimage-config.cmake
2023-08-16T15:47:12.6022448Z 
2023-08-16T15:47:12.6022805Z   but it set freeimage_FOUND to FALSE so package "freeimage" is considered to
2023-08-16T15:47:12.6023299Z   be NOT FOUND.
2023-08-16T15:47:12.6023713Z Call Stack (most recent call first):
2023-08-16T15:47:12.6024195Z   cmake/FindFreeimage.cmake:13 (find_package)
2023-08-16T15:47:12.6024671Z   cmake/SearchForStuff.cmake:6 (include)
2023-08-16T15:47:12.6025135Z   CMakeLists.txt:163 (include)
2023-08-16T15:47:12.6025452Z 
2023-08-16T15:47:12.6025572Z 
2023-08-16T15:47:12.6060283Z --   Found D:/a/gazebo-classic/gazebo-classic/vcpkg/installed/x86-windows-release/include/FreeImage.h
2023-08-16T15:47:14.9965206Z --   Looking for FreeImage.h - found
2023-08-16T15:47:15.0005790Z --   Looking for libfreeimage - found
~~~

There is any specific reason for which we can't just rely on the existing fallback logic?

talregev replied:
I think it better to add it correctly if we want later to compile it with vpckg with linux.
I will try to unset the the same.

traversaro replied:
> I think it better to add it correctly if we want later to compile it with vpckg with linux. I will try to unset the the same.

We can do changes required for other PRs in the corresponding PR. In general, it is ideal for facilitate review that the PR as as small as possible, without unrelated changes, see https://gazebosim.org/docs/all/contributing and https://google.github.io/eng-practices/review/developer/small-cls.html .

talregev replied:
If you insist, I will revert these changes.

talregev replied:
I revert the changes.

talregev committed:
bug fix:
- revert tbb changes
- unset(freeimage_FOUND CACHE)
- unset(tinyxml_FOUND CACHE)
- set(Simbody_LIBRARIES ${Simbody_STATIC_LIBRARIES}) when ${Simbody_LIBRARIES} not found. this is static build case

traversaro commented on cmake/SearchForStuff.cmake:
```diff
@@ -165,8 +165,11 @@ if (PKG_CONFIG_FOUND)
+  if (Simbody_FOUND)  
+    # When simbody is found but it static libs, we need to add it manually.
+    if ("${Simbody_LIBRARIES}" STREQUAL "" OR "${Simbody_LIBRARIES}" STREQUAL "Simbody_LIBRARIES-NOTFOUND")
+    set(Simbody_LIBRARIES ${Simbody_STATIC_LIBRARIES}),  
```
```suggestion
      set(Simbody_LIBRARIES ${Simbody_STATIC_LIBRARIES})
```

Please indent and remove spurious `,` and trailing whitespace, thanks.

talregev replied:
My bad. Thank you for your comments and your fixes.

talregev committed:
Review comments:
- revert changes of freeimages
- revert changes of tinyxml

talregev commented:
@traversaro 
Ready for review.

talregev commented:
@traversaro 
Gentle remind you about this PR.
Also if you can update status or what you do next, it will be nice.
Thank you for your time.

traversaro commented on cmake/SearchForStuff.cmake:
```diff
@@ -679,7 +689,11 @@ endif ()
+if ("${GDAL_LIBRARY}" STREQUAL "")
+  unset(GDAL_FOUND CACHE)
+  include (FindGDAL)
+endif()
```
Is this necessary for the PR to work? Can't we just do `find_package(GDAL CONFIG)` if `include(FindGDAL)` does not find GDAL?

talregev replied:
> Is this necessary for the PR to work? Can't we just do `find_package(GDAL CONFIG)` if `include(FindGDAL)` does not find GDAL?

I think it will not work for this PR. 
Do you want to try it and see?

traversaro replied:
Yes, please. As I already clarified earlier in the review, we try to be conservative, and only add new ways of finding libraries only in the case old ways do not work. If there is a problem in that case, we can debug.

talregev replied:
ok. I am trying.

traversaro replied:
It seems that this version is working.  Given this, are you sure that the:
~~~cmake
if ("${GDAL_LIBRARY}" STREQUAL "")
  unset(GDAL_FOUND CACHE)
  find_package(GDAL CONFIG)
endif()
~~~
is actually necessary?

talregev replied:
I will remove it.
Any other comments?
I will work on it on the evening.

traversaro replied:
Thanks, for the future if we can test if changes are actually necessary before submitting the PR it can be helpful to avoid reducing the time spent in review, thanks!

talregev replied:
> Thanks, for the future if we can test if changes are actually necessary before submitting the PR it can be helpful to avoid reducing the time spent in review, thanks!

I already have that PR with minimal changes. But the repo maintainer didn't respond, So I continue to developing, Also expend to x86 that now is working, also combined it with some Linux fixes. 

So for the future if we have the same response rate like now, I think it will be ok.

If you have more comments for now that you think that something is not necessary, please let me know on that spot that I will fix everything this evening. Thank you! :)

traversaro replied:
> I already have that PR with minimal changes. But the repo maintainer didn't respond, So I continue to developing,

If you want to continue developmenent to add more feature on top of a PR, you can just create a new branch out of the PR and add more work there.

talregev replied:
This is already the new branch with the new PR.
The old branch and old PR not relevant anymore, and I close it for your request.

talregev replied:
done.

talregev committed:
include (FindGDAL) first

talregev committed:
remote find package FindGDAL

talregev commented:
@traversaro Thank you for your comments. I am also agree to minimal and small PRs. 
This is Gentle remind you about this PR.
Anything I can do to move it forward. I think it the minimal changes for this PR.
I will also happy to know the current status of this PR and if it need to have more approvals. 

Thank you for your time and effort. We both will make gazebo better.

talregev commented:
@traversaro Any update?

talregev commented:
@traversaro gentle remind you about this PR.
btw, the problem of windows conda fix itself or this PR fix it by accident?

talregev commented:
@traversaro Can you take a look?

scpeters reviewed (commented):
thank you for your patience with this pull request process. This project is in maintenance mode, so it is not a high priority for us, but I appreciate the work that you have done to demonstrate how to use gazebo-classic on windows with vcpkg.

I think the remaining changes to CMakeLists.txt and cmake/SearchForStuff.cmake don't look harmful, and I don't mind adding the vcpkg metadata if it supports some users.

I looked at the artifact size, and there are 2 uploads of about 500 MB each; I don't think we should upload artifacts of this size

Otherwise, I think we're almost there, pending final thoughts from @traversaro

talregev committed:
Remove upload

talregev commented:
> thank you for your patience with this pull request process. This project is in maintenance mode, so it is not a high priority for us, but I appreciate the work that you have done to demonstrate how to use gazebo-classic on windows with vcpkg.
> 
> I think the remaining changes to CMakeLists.txt and cmake/SearchForStuff.cmake don't look harmful, and I don't mind adding the vcpkg metadata if it supports some users.
> 
> I looked at the artifact size, and there are 2 uploads of about 500 MB each; I don't think we should upload artifacts of this size
> 
> Otherwise, I think we're almost there, pending final thoughts from @traversaro

Thank you for your replay. I remove the upload. Please review again.

traversaro reviewed (approved):
Ok for me, thanks for the PR!

traversaro commented:
@talregev I proposed a more descriptive name for the PR, if you want to change it further feel free to do it!

talregev commented:
> @talregev I proposed a more descriptive name for the PR, if you want to change it further feel free to do it!

That great from my side!
Any thing I can do to move this PR further?

traversaro commented:
> > @talregev I proposed a more descriptive name for the PR, if you want to change it further feel free to do it!
> 
> That great from my side! Any thing I can do to move this PR further?

No, merged!

talregev commented:
> > > @talregev I proposed a more descriptive name for the PR, if you want to change it further feel free to do it!
> > 
> > 
> > That great from my side! Any thing I can do to move this PR further?
> 
> No, merged!

Thank you!
